#### 1. What is the mathematical formula for a linear SVM?
y=w^T+b

#### Q2. What is the objective function of a linear SVM?
- The goal of linear SVM classification is to find the hyperplane that maximizes the margin between two classes. The margin is the distance between the hyperplane and the closest data points from each class. The larger the margin, the better the separation between the classes.

#### Q3. What is the kernel trick in SVM?
- The kernel trick is used to transform the data into a higher-dimensional space where it is easier to separate the classes using a hyperplane. This is done by computing the dot product between the transformed data points, which is equivalent to applying a nonlinear function to the data.

#### Q4. What is the role of support vectors in SVM Explain with example
- A support vector in SVM is a data point that has the largest margin from the decision boundary. These are the points that are closest to the decision boundary and have the most influence on the position of the boundary.

#### Q5. Illustrate with examples and graphs of Hyperplane, Marginal plane, Soft margin and Hard margin in SVM?
- Hard margin SVM does not allow for misclassification, while soft margin SVM allows for some misclassification
- The margin is the distance between the hyperplane and the closest data points from each class.
- the hyperplane that maximizes the margin between two classes

#### Q6. SVM Implementation through Iris dataset.

Bonus task: Implement a linear SVM classifier from scratch using Python and compare its
performance with the scikit-learn implementation.

- Load the iris dataset from the scikit-learn library and split it into a training set and a testing setl
- Train a linear SVM classifier on the training set and predict the labels for the testing setl
- Compute the accuracy of the model on the testing setl
- Plot the decision boundaries of the trained model using two of the featuresl
- Try different values of the regularisation parameter C and see how it affects the performance of the model.

In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [14]:
from sklearn.datasets import load_iris

In [18]:
data=load_iris()

In [20]:
df=pd.DataFrame(data.data, columns=data.feature_names)
df['target']=data.target

In [21]:
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

In [22]:
df['target'].unique()

array([0, 1, 2])

In [24]:
X=df.drop('target',axis=1)
y=df['target']

In [25]:
from sklearn.model_selection import train_test_split

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

## create the SVM

In [27]:
from sklearn.svm import SVC

In [28]:
svc=SVC(kernel='linear')

In [29]:
svc.fit(X_train,y_train)

SVC(kernel='linear')

In [30]:
y_predict=svc.predict(X_test)

In [31]:
## check the accuracy
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report

In [32]:
print(confusion_matrix(y_test,y_predict))

[[19  0  0]
 [ 0 15  0]
 [ 0  0 16]]


In [33]:
print(accuracy_score(y_test,y_predict))

1.0


In [34]:
print(classification_report(y_test,y_predict))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      1.00      1.00        15
           2       1.00      1.00      1.00        16

    accuracy                           1.00        50
   macro avg       1.00      1.00      1.00        50
weighted avg       1.00      1.00      1.00        50



In [35]:
parameter={
    'kernel' : ('linear', 'poly', 'rbf'),
    'C':[0.1,10,100,1000],
    'gamma' : ('scale', 'auto')
}
parameter

{'kernel': ('linear', 'poly', 'rbf'),
 'C': [0.1, 10, 100, 1000],
 'gamma': ('scale', 'auto')}

In [36]:
from sklearn.model_selection import GridSearchCV

In [38]:
gv=GridSearchCV(SVC(),param_grid=parameter,cv=5,verbose=3)

In [39]:
gv.fit(X_train,y_train)

Fitting 5 folds for each of 24 candidates, totalling 120 fits
[CV 1/5] END .C=0.1, gamma=scale, kernel=linear;, score=1.000 total time=   0.0s
[CV 2/5] END .C=0.1, gamma=scale, kernel=linear;, score=0.900 total time=   0.0s
[CV 3/5] END .C=0.1, gamma=scale, kernel=linear;, score=0.900 total time=   0.0s
[CV 4/5] END .C=0.1, gamma=scale, kernel=linear;, score=1.000 total time=   0.0s
[CV 5/5] END .C=0.1, gamma=scale, kernel=linear;, score=0.900 total time=   0.0s
[CV 1/5] END ...C=0.1, gamma=scale, kernel=poly;, score=0.950 total time=   0.0s
[CV 2/5] END ...C=0.1, gamma=scale, kernel=poly;, score=0.900 total time=   0.0s
[CV 3/5] END ...C=0.1, gamma=scale, kernel=poly;, score=0.900 total time=   0.0s
[CV 4/5] END ...C=0.1, gamma=scale, kernel=poly;, score=1.000 total time=   0.0s
[CV 5/5] END ...C=0.1, gamma=scale, kernel=poly;, score=0.950 total time=   0.0s
[CV 1/5] END ....C=0.1, gamma=scale, kernel=rbf;, score=0.800 total time=   0.0s
[CV 2/5] END ....C=0.1, gamma=scale, kernel=rbf

GridSearchCV(cv=5, estimator=SVC(),
             param_grid={'C': [0.1, 10, 100, 1000], 'gamma': ('scale', 'auto'),
                         'kernel': ('linear', 'poly', 'rbf')},
             verbose=3)

In [40]:
gv.best_params_

{'C': 0.1, 'gamma': 'scale', 'kernel': 'linear'}

In [41]:
svc_best=SVC(C=0.1, gamma='scale', kernel= 'rbf')

In [42]:
svc_best.fit(X_train,y_train)

SVC(C=0.1)

In [43]:
y_predbest=svc_best.predict(X_test)

In [44]:
print(classification_report(y_test,y_predbest))
print(confusion_matrix(y_test,y_predbest))
print(accuracy_score(y_test,y_predbest))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       0.94      1.00      0.97        15
           2       1.00      0.94      0.97        16

    accuracy                           0.98        50
   macro avg       0.98      0.98      0.98        50
weighted avg       0.98      0.98      0.98        50

[[19  0  0]
 [ 0 15  0]
 [ 0  1 15]]
0.98


## C: 1.0 to get best acurcay 1
## C :0.1 to get the accurcay 0.98